File: app/schemas/space_item.py
from uuid import UUID
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime

from app.core.enums import Condition, SpaceItemStatus


class SpaceItemCreate(BaseModel):
    """
    Payload for creating expected items in a space.

    This represents the blueprint (not actual physical items).
    """

    space_id: UUID
    item_id: UUID

    # Expected quantity of this item in the space
    quantity: Optional[int] = Field(default=1, ge=1)

    # Whether this item is mandatory for the space
    is_required: Optional[bool] = True

    # Baseline condition expectation (initial state)
    condition: Optional[Condition] = "good"

    # Lifecycle status of expected item
    status: Optional[SpaceItemStatus] = "active"


class SpaceItemUpdate(BaseModel):
    """
    Update payload for space items.
    Allows controlled updates (no structural changes).
    """

    quantity: Optional[int] = Field(default=None, ge=1)
    is_required: Optional[bool] = None
    condition: Optional[Condition] = None
    status: Optional[SpaceItemStatus] = None


class SpaceItemResponse(BaseModel):
    """
    Standard API response model
    """

    id: UUID
    space_id: UUID
    item_id: UUID

    quantity: int
    is_required: bool
    condition: Condition
    status: SpaceItemStatus

    class Config:
        from_attributes = True


class SpaceItemInventoryResponse(SpaceItemResponse):
    item_name: str
    qr_code: str
    last_inspection_id: Optional[UUID] = None
    last_inspection_at: Optional[datetime] = None
    last_inspection_image_url: Optional[str] = None

File: mobile/lib/models/space_item.dart
class SpaceItem {
  final String id;
  final String spaceId;
  final String itemId;
  final String itemName;
  final String categoryName;
  final String qrCode;
  final String? lastInspectionId;
  final DateTime? lastInspectionAt;
  final String? lastInspectionImageUrl;
  final int quantity;
  final bool isRequired;
  final String condition;
  final String status;

  const SpaceItem({
    required this.id,
    required this.spaceId,
    required this.itemId,
    required this.itemName,
    this.categoryName = 'other',
    required this.qrCode,
    required this.lastInspectionId,
    required this.lastInspectionAt,
    required this.lastInspectionImageUrl,
    required this.quantity,
    required this.isRequired,
    required this.condition,
    required this.status,
  });

  factory SpaceItem.fromJson(Map<String, dynamic> json) {
    return SpaceItem(
      id: json['id'].toString(),
      spaceId: json['space_id'].toString(),
      itemId: json['item_id'].toString(),
      itemName: json['item_name']?.toString() ?? 'Unnamed item',
      categoryName: json['category_name']?.toString() ?? 'other',
      qrCode: json['qr_code']?.toString() ?? json['id'].toString(),
      lastInspectionId: json['last_inspection_id']?.toString(),
      lastInspectionAt: _asDateTime(json['last_inspection_at']),
      lastInspectionImageUrl: json['last_inspection_image_url']?.toString(),
      quantity: _asInt(json['quantity']),
      isRequired: json['is_required'] != false,
      condition: json['condition']?.toString() ?? 'unknown',
      status: json['status']?.toString() ?? 'unknown',
    );
  }

  bool get needsAttention {
    return condition != 'good' || status != 'active';
  }
}

int _asInt(Object? value) {
  if (value is int) return value;
  if (value is num) return value.toInt();

  return int.tryParse(value?.toString() ?? '') ?? 0;
}

DateTime? _asDateTime(Object? value) {
  if (value == null) return null;

  return DateTime.tryParse(value.toString());
}
